rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
 type TValEnter<P, T> = TraceEnter<TValAgent<P>, T>;
 type TKeyEnter<P, T> = TraceEnter<TKeyAgent<P>, T>;
 
-/// Diff associated with records in differential dataflow
+#[derive(
+    Abomonation, Copy, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone, Serialize, Deserialize,
+)]
+pub struct CheckedWeight {
+    pub value: i32,
+}
+
+/// CheckedWeight trait implementations
+
+impl Semigroup for CheckedWeight {
+    fn is_zero(&self) -> bool {
+        self.value == 0
+    }
+}
+
+impl<'a> AddAssign<&'a Self> for CheckedWeight {
+    fn add_assign(&mut self, other: &'a Self) {
+        // intentional panic on overflow
+        self.value = self
+            .value
+            .checked_add(other.value)
+            .expect("Weight overflow");
+    }
+}
+
+impl Add for CheckedWeight {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: self
+                .value
+                .checked_add(other.value)
+                .expect("Weight overflow"),
+        }
+    }
+}
+
+impl Mul for CheckedWeight {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: self.value.checked_mul(rhs.value).expect("Weight overflow"),
+        }
+    }
+}
+
+impl Neg for CheckedWeight {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: 0_i32.checked_sub(self.value).expect("Weight overflow"),
```

Comment:
This is redundant, negation can't fail since the positive and negative integer spaces are equal. Any positive number can be converted into a negative one

Reply:
They are not really equal, but indeed the negative one is bigger than the positive one.

Reply:
But this check is actually necessary, in case that self.value is MIN_INT: https://towardsdatascience.com/unsinged-signed-integers-and-casting-in-rust-9a847bfc398f#3e54

Reply:
Why not use `checked_neg`?